### PR TITLE
[SPARK-12877] [ML] Add train-validation-split to pyspark

### PR DESCRIPTION
**Pull request by JeremyNixon**
## What changes were proposed in this pull request?

The changes proposed were to add train-validation-split to pyspark.ml.tuning.
## How was the this patch tested?

This patch was tested through unit tests located in pyspark/ml/test.py.

This is my original work and I license it to Spark.

### CODE REVIEW

**Comment by mengxr** [comment]
ok to test

File: python/pyspark/ml/tuning.py [hunk 1]
```diff
@@ -288,6 +289,172 @@ def copy(self, extra=None):
         return CrossValidatorModel(self.bestModel.copy(extra))
 
 
+class TrainValidationSplit(Estimator, HasSeed):
```
Comment (sethah) [inline python/pyspark/ml/tuning.py]: Missing docstring and doctests.

File: python/pyspark/ml/tuning.py [hunk 2]
```diff
@@ -288,6 +289,172 @@ def copy(self, extra=None):
...
+        Creates a copy of this instance with a randomly generated uid
+        and some extra params. This copies creates a deep copy of
+        the embedded paramMap, and copies the embedded and extra parameters over.
+
+        :param extra: Extra parameters to copy to the new instance
+        :return: Copy of this instance
+        """
+        if extra is None:
+            extra = dict()
+        newCV = Params.copy(self, extra)
```
Comment (sethah) [inline python/pyspark/ml/tuning.py]: Should this be called `newTVS`?

File: python/pyspark/ml/tuning.py [hunk 3]
```diff
@@ -288,6 +289,172 @@ def copy(self, extra=None):
...
 
 
+class TrainValidationSplit(Estimator, HasSeed):
+
+    estimator = Param(Params._dummy(), "estimator", "estimator to be tested")
+    estimatorParamMaps = Param(Params._dummy(), "estimatorParamMaps", "estimator param maps")
+    evaluator = Param(
+        Params._dummy(), "evaluator",
+        "evaluator used to select hyper-parameters that maximize the metric")
+    trainRatio = Param(Params._dummy(), "trainRatio", "proportion for train-validation ratio")
```
Comment (sethah) [inline python/pyspark/ml/tuning.py]: The description for this parameter in scala is "ratio between training set and validation set (>= 0 && <= 1)". It would be good to maintain consistency here.

File: python/pyspark/ml/tuning.py [hunk 4]
```diff
@@ -288,6 +289,172 @@ def copy(self, extra=None):
         return CrossValidatorModel(self.bestModel.copy(extra))
 
 
+class TrainValidationSplit(Estimator, HasSeed):
+
+    estimator = Param(Params._dummy(), "estimator", "estimator to be tested")
+    estimatorParamMaps = Param(Params._dummy(), "estimatorParamMaps", "estimator param maps")
+    evaluator = Param(
+        Params._dummy(), "evaluator",
+        "evaluator used to select hyper-parameters that maximize the metric")
```
Comment (sethah) [inline python/pyspark/ml/tuning.py]: nit: "maximize the metric" -> "maximize the validated metric"

File: python/pyspark/ml/tuning.py (hunk 3 above)
Comment (JeremyNixon) [inline python/pyspark/ml/tuning.py]: The Scala parameter comments above trainRatio also say 'Param for ratio between train and validation data. Must be between 0 and 1.' - seems faster to parse for a developer. Have a preference?

File: python/pyspark/ml/tuning.py (hunk 1 above)
Comment (JeremyNixon) [inline python/pyspark/ml/tuning.py]: Updated.

File: python/pyspark/ml/tuning.py (hunk 2 above)
Comment (JeremyNixon) [inline python/pyspark/ml/tuning.py]: Updated.

File: python/pyspark/ml/tuning.py (hunk 4 above)
Comment (JeremyNixon) [inline python/pyspark/ml/tuning.py]: Updated.

File: python/pyspark/ml/tuning.py (hunk 3 above)
Comment (sethah) [inline python/pyspark/ml/tuning.py]: Sounds good.

File: python/pyspark/ml/tuning.py [hunk 5]
```diff
@@ -288,6 +289,195 @@ def copy(self, extra=None):
         return CrossValidatorModel(self.bestModel.copy(extra))
 
 
+class TrainValidationSplit(Estimator, HasSeed):
+    """
+    Train-Validation-Split.
```
Comment (sethah) [inline python/pyspark/ml/tuning.py]: There needs to be a space here or the docs won't generate correctly. You can check this on your local machine by running `jekyll build` in the docs folder.

File: python/pyspark/ml/tuning.py [hunk 6]
```diff
@@ -288,6 +289,195 @@ def copy(self, extra=None):
...
+        self._setDefault(trainRatio=0.75)
+        kwargs = self.__init__._input_kwargs
+        self._set(**kwargs)
+
+    @since("2.0.0")
+    @keyword_only
+    def setParams(self, estimator=None, estimatorParamMaps=None, evaluator=None, trainRatio=0.75,
+                  seed=None):
+        """
+        setParams(self, estimator=None, estimatorParamMaps=None, evaluator=None, trainRatio= 0.75,\
```
Comment (sethah) [inline python/pyspark/ml/tuning.py]: nit: remove space after `trainRatio=`

**Comment by sethah** [comment]
LGTM other than minor comments. Perhaps @yanboliang or @holdenk could have a look?

File: python/pyspark/ml/tuning.py (hunk 6 above)
Comment (JeremyNixon) [inline python/pyspark/ml/tuning.py]: Updated.

File: python/pyspark/ml/tuning.py (hunk 5 above)
Comment (JeremyNixon) [inline python/pyspark/ml/tuning.py]: Updated.

**Comment by mengxr** [comment]
Merged into master. Thanks!

**Comment by mengxr** [comment]
@JeremyNixon I don't know your JIRA account. Could you please leave a comment on the JIRA page? So I can set you as the assignee. Thanks!

**Comment by JeremyNixon** [comment]
@mengxr Done, no problem!
